### DISCUSSION TITLE
GAT for Time Series ( like a transformer for nlp )

### DISCUSSION THREAD

**Discussion by claCase** [danielegrattarola/spektral#304] [Q&A] [answered]
Hi @danielegrattarola , 
I'd like to use GAT for time series where nodes are sampled points and edges the dependeces between points. So the input X would be of shape [batch, seq_len, feature_dim]. Theoretically one would just need to use a lower triangular adjacency matrix of shape [batch, seq_len, seq_len] to implement a causal temporal masking. However the model surprisingly doesn't really work. Any suggestion why would that be the case? 

I've written a test that can be found here [TEST](https://github.com/claCase/Master-Thesis/blob/master/src/Tests/test_sequential_attn.py)
which implements a GAT model with batched inputs. The function "generate_list_lower_triang" implements a list of lower triangular matrices for each trajectory with lags up to "lag" so that the attention gets computed upto lag "lag".  

EDIT: 
Setting the lag to 3 makes the model learn better. Here is a comparison with an LSTM loss curve
![loss_gat_lstm](https://user-images.githubusercontent.com/36463823/142192167-7c249e0b-28d5-4141-9eaa-d71b04f160cd.png)

**Comment by danielegrattarola** [comment, marked as answer]
Hi, 

what is the expected result? Do you see meaningful differences with a typical transformer?
I have looked at the code and I don't really see any issues, except that maybe 2 channels and 1 head are not enough neurons for GAT? The LSTM has 10...

**Reply by claCase** [reply]
Hi @danielegrattarola,
Thanks for the answer!
You're right, increasing the channels to 10 and heads to 2 and then having a final output dense layer makes the GAT train faster than the LSTM and achieves almost the same loss.
![loss_gat_lstm_10h_2h_1lag](https://user-images.githubusercontent.com/36463823/142583645-3a073c05-86bb-44eb-855f-ced2c3085892.png)

 I think the problem was that I was initially setting the lag too high, which doesn't make sense for this dataset since it is just a moving average process/random walk. I though the model would automatically learn to ignore the older lags and focus more on the recent ones.
